### DISCUSSION TITLE
Grid view improvements

### DISCUSSION THREAD

**Discussion by trappar** [SynthstromAudible/DelugeFirmware#365] [Ideas]
Starting a discussion so we can organize feedback/ideas/requests related to this new feature.

**Stuff that was suggested which has been implemented**
* Launching multiple clips at once
* Ability to create audio clips more quickly
* Track colors
* One touch clip creation / editing
* Fixed vertical scroll direction

**Comment by trappar** [comment]


**Reply by PaulFreund** [reply]
Agreed, I will most likely change cloning to long press instead to allow for this, maybe with flashing the selected source clip. Thoughts?

**Reply by trappar** [reply]
Seems like it could be simpler than that. What holding a clip and pressing another pad should do should depend on the destination pad.
* When there is no clip in the destination pad, clone the clip.
* When there is a clip in the destination pad, toggle the play queue state for both pads.

Maybe I'd have to play around with both options to really know which would be better. Also, I think you should be able to press any number of pads simultaneously and have them all toggle play states, not just two.

**Reply by PaulFreund** [reply]
Please check #500, Launch mode can arm or toggle multiple clips with that

**Comment by trappar** [comment]
**Moving clips**

There's no current way to move a clip in grid mode other than copying it and deleting the previous clip. Moving clips could follow the existing UX for moving a kit row. Press one clip, then press the destination clip as if to copy, then while holding both pads hit the vertical encoder.

**Reply by PaulFreund** [reply]
I was thinking about using shift instead of the vertical encoder but for clips that are far away it might be really awkward to figure out the right hand positions :D

**Reply by sichtbeton** [reply]
its needed, agreed.

why not jjust hold "clip pad" + turn vertical encoder up and down?

**Reply by chewabledrapery** [reply]
Again I agree with @sichtbeton on this, hold pad, dial vertical encoder, let go, clip moved...

**Reply by darenager** [reply]
I agree hold clip + turn vertical would be great!

**Reply by adam-p** [reply]
Idiosyncratic data point: While messing with grid view for the first time, I naively tried to reorder rows (sections) by holding the coloured audition pad and turning the vertical encoder. I also tried by holding an arbitrary clip pad in the row. (Similarly for reordering instruments: Hold one of the clip pads in a column and turn the horizontal encoder.)

I could certainly cope with more presses than that, though. If I hadn't been messing around with a song created in row view (which had different section ordering), then there's much less chance that I'd be reordering.

**Comment by m-m-adams** [comment]
I really want to be able to create a clip by just clicking a square. Having to use a button combo is a noticeable loss compared to the default row view for live looping

**Reply by PaulFreund** [reply]
Fully agree, I will find a way to make it feel natural :)

**Reply by sichtbeton** [reply]
Just clicking a pad and either “kit” / “MIDI” / “synth “ would be natural for me.

**Reply by sichtbeton** [reply]
And clicking “select “ for an audio clip also.

**Reply by PaulFreund** [reply]
Please check #500, in Edit mode no additional presses are required

**Reply by sichtbeton** [reply]
Nice!

**Comment by mcabrerapf** [comment]
Seems to me that vertical scroll works opposite to how it works on all other views (clip, arranger, row)

In GRID view if you turn the vertical scroll knob clockwise it will scroll down, if you do this in a CLIP view it scrolls up.

**Reply by trappar** [reply]
Good catch!

**Reply by PaulFreund** [reply]
I knew it felt off but I wasn't quite sure. Fixed in #431

**Reply by PaulFreund** [reply]
Would it be okay to delete discussion items that are implemented?

**Comment by sichtbeton** [comment]
Track colors (can be inherited from the usual row mode) would be awesome, because right now it’s rather hard to know what’s what. Also it  would help users of the standard view transition easier to the new grid view.

**Reply by jamiefaye** [reply]
I came here to make the same suggestion. It may also let people see what is happening if you have multiple clips with the same row color. You can then cycle between clips each time you trigger, like we do on the Song view.

**Reply by sichtbeton** [reply]
I think without the orientation that track colors give, a not so small number of users will probably be plainly confused and discard the actually great new feature.

**Reply by TCPderp** [reply]
Much needed QOL improvement. I REALLY like the new grid mode, but was distracted by the colors. Needs to corollate with row mode.

**Comment by seangoodvibes** [comment]
**Split Grid View Mode**

As discussed, in Discord, I am documenting this new layout that will build upon the existing Grid View but also brings in elements from the existing Song View and the Keyboard View.

Specifications:

- The 8 x 16 Deluge Grid will be split into two 8 x 8 sections

<img width="532" alt="Screen Shot 2023-09-01 at 8 51 55 PM" src="https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/28df6fcd-596c-4d2b-b408-539a982e25df">

- The left 8 x 8 side will including a **Horizontal** grid view layout whereby each row is a section and each column is a clip within that section. Additional functionality should be incorporated into the left side grid view:
  - punch in / punch out audio clip recording (press a pad in a section and it will automatically start recording the audio coming into the deluge into an audio clip. press the pad again to stop the recording so that the clip can begin looping the recorded audio). when a pad is "punched in" it will flash to indicate you are recording audio into it. 
    - for new audio clips, the play cursor on the right side would be red to indicate its recording a fresh clip and the length has not been set yet.
  - punch in / punch out overdubbing of existing audio clip. 
    - for overdubbed audio clips, the play cursor on the right side would be white to indicate its overdubbing onto a clip for which the length has been set.
  
- The right 8 x 8 side will include a **Horizontal** preview of the underlying section's contents for the selected clip in that section. This would be displayed in the same way that the original song view displays a clips contents. Additional functionality should be incorporated into the right side grid view:
  - show the play cursor so you can track where the play position is at in each active clip.
  - substitute the preview with the keyboard view by pressing the keyboard button. You could have multiple keyboard keyboard modes:  
     1) a single row keyboard for each selected clip in each section; 
     2) a 8x8 keyboard for one specific grid clip 
     3) I could also see other keyboard / drum variants - e.g. 2x8, 4x8. it would almost be like a performance layout for keyboard/drums whereby you could choose clips and setup the right side as you want (eg one clip is set with a 1x8 drum layout, another clip is set with a 2x8 piano layout, another with a 5x8 in key layout).

- The first column of the sidebar would be used to start/stop sections. When starting a section, it will start with the previously selected grid clip in a section. or if that grid clip was deleted since the previous playback, it will select the first grid clip in a section.

- The second column of the sidebar could be used for different performance layouts (eg like the custom keyboard view layout mentioned above).

- There should be horizontal / vertical scrolling on both sides of the view (eg to see more sections, or to set the notes/sounds in focus for the keyboard / drum layout)

**Reply by sichtbeton** [reply]
This would be a variation on the Split View found in the Ableton push. Very nice idea.

**Comment by mark0n3** [comment]
Much like in ableton grid mode will likely be my default view, thanks so much for implementing it. It would be great to be able to name grid tracks, clips and and scenes directly in grid mode.

**Comment by sichtbeton** [comment]
While i get why the two modes are there, i would propose to get rid of them. its just too fiddly and you alway forget in which mode you are.

- no modes 
- always in playback (formerly green) mode
- a clip will play on release of the pad and not on "pad press", the timing difference probably won't matter in real life
- holding a pad and turning an encoder for parameters doesn't trigger / untrigger clip on release
- "enter" a clip by holding any of the audition buttons + a clip pad
- hitting an empty pad either (optionally) creates a new clip or stops all clips in that track (like Ableton)

**Reply by TCPderp** [reply]
Was just going to comment a similar train of thought. 

- No modes.
- Always in playback 
- Play clip on release. (_I don't care about this differentiation, but I can see how this could be useful. easy to hold multiple clips and have them queue together._)
- holding a pad and turning an encoder for parameters doesn't trigger / untrigger clip on release (_Didn't notice this, but if it's true this is not desired_)
- **How about enter a clip by double tapping it.**
- Agree to stopping track if there is no clip. 

Just one small change to your thoughts.

**Comment by sichtbeton** [comment]
moving a track column: currently you can't move a track and thus rearrange the grid. proposed workflow:

- hold any track column pad that has a clip in it
- press and turn the horizontal encoder to move the track column left <> right
- let go of the encoder and pad

**Reply by chewabledrapery** [reply]
Agree, I LOVE the grid view but i really need to be able to re-arrange the columns/instruments, even it were to inherit the order from row mode it would work (you could switch in and out of views?) but it seems to arbitrarily organise the columns @sichtbeton suggestion is exactly what i was hoping to suggest. Excellent work and thank you so much for this it really is a game changer!

**Reply by m-m-adams** [reply]
It copies the order from arranger so you should be able to rearrange there for now

**Reply by sichtbeton** [reply]
Oh that’s good to know!

**Reply by davidvandenbor** [reply]
This doesn't work if the clips are midi clips! Question: I can't for the life of me seem to change the **horizontal order** (clip position) of the midi clips I made in clip mode. Is this a bug, or is this feature just not implemented? Isn't this basic functionality that should be simple to do on the Deluge? Kind regards, David!

**Reply by seangoodvibes** [reply]
> This doesn't work if the clips are midi clips! Question: I can't for the life of me seem to change the **horizontal order** (clip position) of the midi clips I made in clip mode. Is this a bug, or is this feature just not implemented? Isn't this basic functionality that should be simple to do on the Deluge? Kind regards, David!

This was not implemented for grid view

The work around is to go into arranger view and re order tracks and it will be reflected in grid view

**Comment by sichtbeton** [comment]
also, and i know i will infuriate @PaulFreund: having only one section color active at first in the grid would be way cleaner when starting out. when you add new clips to a new section, the new section color comes on.

it tidies up the screen and is clean, efficient and to the point: show me what's actually working. 

but as I said, we already discussed this at length in the discord, so this here is just to push him over the edge (haha).

**Comment by mark0n3** [comment]
Add FX performance and record mode. 

A third option to green and blue, where the left sidebar buttons each represent an effect. Hold the effect to adjust its parameters on the gold buttons, hold an effect and a clip or multiple clips to apply the effect to those clips. If record is on, record the result into a new respective row for each clip held. 

Allow fx buttons to be midi learned for easier control from another device or software (eg ios) control surface. 

Add the option to record effects into effects-only clips such that any clips played at the same time have those effects applied to them.

**Comment by trappar** [comment]
**Add track colors to rows mode**

A community option can be added to switch between default behavior vs. using track colors for this column. The arm states would just reuse the way that track colors are represented in grid mode.

**Comment by trappar** [comment]
**Stop clips by clicking empty clip**

In green mode, add ability to queue up stopping the current clip for a column by pressing an empty clip in the same column. Personally I find pressing the active clip a bit unintuitive and I tend to reach for an empty clip (probably muscle memory from Push)

**Reply by trappar** [reply]
This has been implemented.

**Comment by trappar** [comment]
**Add mixer mode to grid view**

Add a new mode to the grid mode sidebar for switching to a mixer view. In this mode, each column shows a single parameter's value for the corresponding track by illuminating some number of pads from bottom to top corresponding to the parameter's value. When switching to this mode, it would initially start by showing the volume for each track, and then could be switched by either rotating a gold knob or by pressing a shortcut (shift + pad).

IMO this would be a great way to visualize multiple tracks values relative to each-other, which is currently not something that can be done on Deluge.

Potential extras for this mode:
* Allow the user to hold a grid pad and then rotate one of the knobs (not sure which... maybe select?) to fine-tune a parameter.
* The mode should remember the last parameter edited and recall that parameter upon switching back to this mode after changing away from volume.

**Comment by betonizm** [comment]
Why is there only 12 lines in gridmode? how to create more

**Reply by sapphire-arches** [reply]
The 12 rows correspond to the existing 12 sections you can access in rows mode. Lifting this restriction requires some architectural work and UI design that hasn't been done yet.

**Comment by donluca** [comment]
Would it be possible to have an option so that the grid is "rotated"?

I'd rather have 8 tracks with 16 clips each ready at my fingertips than 16 tracks with 8 clips each.
The buttons on the right can be used to either mute or solo the track (ie: entire row) in this mode
